### DISCUSSION TITLE
Use fortls from a vim + YCM environment?

### DISCUSSION THREAD

**Discussion by ccoustet** [fortran-lang/fortls#103] [Q&A]
Hi,
I mostly code in C. I successfully configured clangd / YouCompleteMe / vim and benefit from all the clangd features, as expected.
As I sometimes code in Fortran, I wonder how I could make use of fortls, keeping all that works for C unchanged (keep using vim VS neovim, keep using YCM VS other completion plugins) ?
Does anyone there have experience/some success using the vim + YCM + fortls combo?
Thanks for any help.

**Comment by gnikit** [comment]
Hi @ccoustet, interesting problem. I am not familiar with YCM, but if I had to take a guess I would say that the Language Server Protocol (LSP) requests should take precedence over any other form of autocomplete, or at least have the same "weight".

What happens if you use YCM + fortls? 
Or is that currently not possible?
If not what problems are you experiencing?

P.S. Configure instructions for [Vim](https://gnikit.github.io/fortls/editor_integration.html#vim-neovim-gvim)

**Comment by gnikit** [comment]
I also found this for configuring YCM with Language Servers https://github.com/ycm-core/YouCompleteMe#lsp-configuration
not sure if it would work

**Reply by ccoustet** [reply]
Hi gnikit,
Sorry, I didn't give you much context on my settings.
I installed fortls using pip.
If I open a Fortan file in vim, fortls starts and ps can see it running.
I added a few lines to my .vimrc too:

```
let g:ycm_language_server += [
  \   { 'name': 'fortran',
  \     'filetypes': [ 'fortran' ],
  \     'cmdline': [ 'fortls --hover_signature --use_signature_help --incremental_sync' ],
  \   }, 
  \ ]
```
When in a Fortran file, some of the YCM tricks work as expected (e.g. :YcmCompleter GoTo to go to a definition), but most don't, like functions signatures, symbols search, symbol renaming, or real-time error checking.
As most of them are listed as fortls features, I guess that the problem comes from my configuration.

Thanks for your help.
~                                                                                                                                                                                        
~                                                                                                                                                                                        
~                                                                                                                                                                                        
~                                                                                                                                                                                        
~

**Reply by gnikit** [reply]
Sorry for not getting back to you sooner. I have been trying to untangle the mess that is vim and neo-vim. I quickly realised that the options I copied into the `fortls` documentation were not correct.

I will post a full update later, but here is where I am at:

1. Vim does not support the Language Server Protocol (LSP) so you need to install either [LanguageClient-neovim](https://github.com/autozimu/LanguageClient-neovim) or [coc.nvim](https://github.com/neoclide/coc.nvim)
2. NeoVim < 0.5.0 does not support LSP so you need to install one the same plugins
3. NeoVim > 0.5.0 supports LSP which can be enabled by installing the default extension https://github.com/neovim/nvim-lspconfig

For more info about NeoVim and LSP support please refer to: https://neovim.io/doc/lsp/

## YCM + fortls

My understanding is that YCM, much like LanguageClient-neovim and coc.nvim **also** has the ability to act as an LSP client, which means that you wouldn't need to use LanguageClient-neovim at all. This is the relevant piece of documentation I found on their webpage: https://github.com/ycm-core/YouCompleteMe#plugging-an-arbitrary-lsp-server

I believe something like this should work, I will try it myself too but let me know as well:

```vimrc
let g:ycm_language_server = 
  \ [ 
  \   {
  \     'name': 'fortran',
  \     'cmdline': [ 'fortls', '--notify_init', '--hover_signature', '--hover_language', 'fortran', '--use_signature_help' ],
  \     'filetypes': [ 'fortran' ],
  \     'project_root_files': [ '.fortls' ]
  \   },
  \ ]
````

I was able to get the go-to to work, similar to you.
I was also able to get the symbol search to work, see figure below with

```vimrc
nmap <leader>yfw <Plug>(YCMFindSymbolInWorkspace)
nmap <leader>yfd <Plug>(YCMFindSymbolInDocument)
```

![image](https://user-images.githubusercontent.com/16143716/168333932-34e357eb-ea8e-41d7-aba4-f93c158791f7.png)


Hover messages

![image](https://user-images.githubusercontent.com/16143716/168334354-9517690b-fb94-4606-aa75-b4665c3c343c.png)

**Comment by ccoustet** [comment]
Thanks for this answer!
Most features are now working: find in workspace, goto definition, refactor rename. Hover popup also works to some extent.
For the old f77 code I am editing just now, the signature popup does not include type information :
![Capture d’écran de 2022-05-16 11-44-42](https://user-images.githubusercontent.com/6315590/168567162-e9f403bc-59d5-47fe-b56d-8cdfe02f7943.png)
Also :YcmCompleter GetDoc calls end in error, as well as GoToCallers and a few others :
```
ValueError: Supported commands are:
GetHover
GoTo
GoToDeclaration
GoToDefinition
GoToDocumentOutline
GoToImplementation
GoToReferences
GoToSymbol
RefactorRename
RestartServer
See the docs for information on what they do.
```
Finally, if I introduce an error in the source code, no real time diagnostic mark appears, nor :YcmForceCompileAndDiagnostics reports anything (output is: No warnings or errors detected.).
Anyway it is a big step forward.

**Reply by gnikit** [reply]
>  GetDoc calls end in error, as well as GoToCallers and a few others :

As YCM reported these commands are not supported. The supported LSP commands are shown in the README. Feel free to open a PR and have a go at implementing them but keep in mind a couple of things:

1) some of this LSP functinoality was added in the latest release of LSP 3.16 so a lot of editors might not even support it 
2) a lot of the Vim extensions for LSP are implementing additional LSP requests that are not part of the standard. The official LSP requests can be found here: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/

> if I introduce an error in the source code, no real time diagnostic mark appears

Diagnostics are served on Document Save, might change that in the future but this was originally done for performance, since serving diagnostics requires the reparsing of the file.

> For the old f77 code I am editing just now, the signature popup does not include type information :

As far as I can tell, that is not a hover message but a signature help, i.e. it completes the function signature while you are typing it. From the looks of it this is missing a lot of info which makes me suspicious. I am not entirely convinced that this pop-up message is served by `fortls`, especially since it looks identical with the `voigt(x,y,v)` definition that you have a few lines after.

A way to test if `fortls` is working would be through the debug interface. From a command line type this


```sh
fortls --debug_filepath your_file.f --debug_rootpath . --debug_hover --debug_line line_num --debug_char column_position
```

You might want to try `--debug_signature` as well.

Please post the results.

**Reply by ccoustet** [reply]
The GetDoc  and GoToCallers are genuine YCM calls that work for C (clangd). I have no idea if they are supported YCM-side for fortran. I plan to contact the YCM team once I fully understand my fortls settings.

The signature help seems to be inspired by the voigt function's code (x,y,v are the names of the parameters in voigt definition). If I rename the function's parameters the signature popup uses the new parameters names.

```
$ fortls --debug_filepath transfer.f --debug_rootpath . --debug_hover --debug_line 344 --debug_char 10

Testing "initialize" request:
  Root = "."
[INFO - 16:02:59] fortls - Fortran Language Server 2.5.0 Initialized
  Successful!

  Source directories:
    /home/chris/Documents/inv_q_molec_haze/src

Testing "textDocument/hover" request:
  File = "transfer.f"
  Line = 344
  Char = 10

  Result:
=======
[{'language': 'fortran90', 'value': 'FUNCTION voigt(x, y, v) RESULT(voigt)'}]
=======
```
```
$ fortls --debug_filepath transfer.f --debug_rootpath . --debug_signature --debug_line 344 --debug_char 10

Testing "initialize" request:
  Root = "."
[INFO - 16:05:13] fortls - Fortran Language Server 2.5.0 Initialized
  Successful!

  Source directories:
    /home/chris/Documents/inv_q_molec_haze/src

Testing "textDocument/signatureHelp" request:
  File = "transfer.f"
  Line = 344
  Char = 10

  No Results!
```
I still cannot have diagnostics, even on save: adding a broken line "vO=voigt(" in the file and saving do nothing in term of diags.
```
$ head transfer.f
      v0=voigt(
      subroutine transfer(nlev,nlay,qn2,qh2,qar,pl,tl,ml,gl,qch4,ql,
     *z,p,ap,t,nview,nsec,sec,lay_min,ncorps,corps,numcor,mass,
     *erot,et,at,freq1,freq2,pas_mult,pas_sort,file_out,iconv,res,
     *itrans,file_trans,file_spec_nn,file_spec_nc,file_spec_cc,
     *file_spec_nh,icloud,wref,taucl,file_sca,
     *file_sp,smin,ttest,nlor,alor,glor,elor,g2lor,nvib,vib,ndeg,
     *n_k,corps_k,icorps_k,nx,freq_x,nsec_x,isec_x,matrix_k,rad,iter)
      implicit double precision (a-h,o-z)
      parameter(nlev_max=96,nlay_max=nlev_max-1,nond_max=200,nsec_max=15,
```
```
$ fortls --debug_filepath transfer.f --debug_rootpath . --debug_diagnostics

Testing "initialize" request:
  Root = "."
[INFO - 16:12:59] fortls - Fortran Language Server 2.5.0 Initialized
  Successful!

  Source directories:
    /home/chris/Documents/inv_q_molec_haze/src

Testing "textDocument/publishDiagnostics" notification:
  File = "transfer.f"

No errors or warnings
```
```
$ make
gfortran-11 -O3 -march=native -flto  -g3 -Wall -Wno-unused-dummy-argument -ffixed-line-length-none -Jobj -c src/voigt.f -o obj/voigt.o
src/transfer.f:1:15:

    1 |       v0=voigt(
      |               1
Error: Syntax error in argument list at (1)
```
Thanks for your time.

**Reply by gnikit** [reply]
## YCM

> The GetDoc and GoToCallers are genuine YCM calls that work for C (clangd). I have no idea if they are supported YCM-side for fortran.

so YCM is a Language Server Protocol (LSP) client. For Vim and neovim there are a few available like CoC, LanguageClient-neovim, etc. `fortls` is a an LSP server. Both the client and the server have to implement the parts of the LSP that are applicable to them i.e. requests and responses. 

I don't know what request GoToCallers and GetDoc are supposed to be aliasing but their names do not match any part of the standard: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/

I would guess that the gotocallers is similar to `textDocument/references` which is well supported and the getdoc fetches the doc part of the hover message, which will not work since it assumes that the hover message is  expressed using `MarkupContent`. For more info see https://github.com/gnikit/fortls/issues/45

Also the supported docs are limited to FORD and Doxygen (and given that both are highly configurable `fortls` supports only the default settings).


## Diagnostics

The diagnostics provided by the server are limited, please read the documentation: https://gnikit.github.io/fortls/README.html#features .
Compile-time diagnostics are usually provided by linters not language servers, since a substantially more complex parser is needed. We have that functionality in VS Code https://github.com/fortran-lang/vscode-fortran-support, but I don't personally know a lot of other Fortran code editors and extensions that support both linting and LSP servers.


## Hover

The hover still doesn't look right (I think!). If you look at the picture I added a few replies back the hover message is supposed to contain the types of the argument list. Do you think you can provide a [Minimal Working Example](https://en.wikipedia.org/wiki/Minimal_reproducible_example) that replicates the hover problem?

**Comment by ccoustet** [comment]
From YCM documentation, 
```
The GoToCallers and GoToCallees subcommands

Populate the quickfix list with the callers, or callees respectively, of the function associated with the current cursor position. The semantics of this differ depending on the filetype and language server.
Only supported for LSP servers which provide the callHierarchyProvider capability.
```

From your answer, I realize that not all the possible errors are diagnosed. A few command line tests show that a non-closed do is detected, while missing args - e.g. voigt() - remains unreported (no warnings or errors detected). However I still need to understand how to configure YCM to provide fortls with the expected argument list, so that it works from vim.

From the previous point I can imagine that YCM doesn't know how to feed fortls calls with proper arguments regarding hover, and that what is displayed is standard vim text analysis. However, I tried to write a tiny example and call fortls in command line mode: 

```
$ cat -n none.f
     1	      function fff(x,y)
     2	      implicit none
     3	      real fff, x,y
     4	      fff = x+y
     5	      return
     6	      end
     7	
     8	      program main
     9	      implicit none
    10	      real y, fff
    11	      y = fff(1.2,3.3)
    12	      write(*,*) y
    13	      end


 $ fortls --debug_filepath none.f --debug_rootpath . --debug_hover --debug_line 11 --debug_char 11

Testing "initialize" request:
  Root = "."
[INFO - 14:08:21] fortls - Fortran Language Server 2.5.0 Initialized
  Successful!

  Source directories:
    /home/chris/Documents/fortl_test

Testing "textDocument/hover" request:
  File = "none.f"
  Line = 11
  Char = 12

  Result:
=======
[{'language': 'fortran90', 'value': 'REAL'}]
=======

```
Here real is the actual type of fff(1.2,3.3). However I would have expected something like: real x real -> real. Or may I have missed something in the way I test?

**Reply by gnikit** [reply]
## YCM docs

`callHierarchyProvider` is not provided, probably will be in the future though.



## Hover

You actually stumbled onto a bug (more like syntax that I was unaware). fortls sees `fff` inside main and thinks it's just the real variable defined and knows nothing about the globally? defined `fff`. That's an oversight on my end since I code mostly in Modern Fortran using OOP. I will file a bug report.

Here is the parser output for fortls:


```log
Testing parser
  File = "test_fixed.f"
  Detected format: fixed

=========
Parser Output
=========

=== No PreProc ===

      function fff(x,y) !!! FUNCTION - Ln:1
          implicit none !!! IMPLICIT - Ln:2
          real fff, x,y !!! VARIABLE - Ln:3
      end !!! END  Scope - Ln:6
      program main !!! PROGRAM - Ln:8
          implicit none !!! IMPLICIT - Ln:9
          real y, fff !!! VARIABLE - Ln:10
      end !!! END  Scope - Ln:13

=========
Object Tree
=========

3: fff
  6: fff::fff
  6: fff::x
  6: fff::y
1: main
  6: main::y
  6: main::fff

=========
Exportable Objects
=========

3: fff
1: main
````
